lib/src/modules/apply.cc

```diff
@@ -0,0 +1,252 @@
+
+namespace PXPAgent {
+namespace Modules {
+
+    static const std::string apply_ACTION { "apply" };
+    static const std::string apply_ACTION_INPUT_SCHEMA { R"(
+    {
+        "type": "object",
+        "properties": {
+            "environment": {
+                "type": "string"
+            },
+            "version": {
+                "type": "string"
+            },
+            "catalog": {
+                "type": "object"
+            },
+            "_noop": {
+                "type": "boolean"
+            },
+            "apply_settings": {
+                "type": "object"
+            }
+        },
+        "required": ["environment", "version", "catalog"]
+    }
+    )" };
+
+    // TODO: Actually manage this file in packaging (or perhaps fetch it from puppetserver).
+    // For now just copy it to the cache dir each time unless there is already a copy
+    static const std::string RUBY_SHIM {
+R"(#! /opt/puppetlabs/puppet/bin/ruby
+# frozen_string_literal: true
+
+require 'fileutils'
+require 'json'
+require 'puppet'
+require 'puppet/configurer'
+require 'securerandom'
+require 'tempfile'
+
+## TODO: Option to read from a file is for debugging. Only read from stdin in production.
+args = JSON.parse(ARGV[0] ? File.read(ARGV[0]) : STDIN.read)
+apply_opts = args['apply_options'] || {}
+# Create temporary directories for all core Puppet settings so we don't clobber
+# existing state or read from puppet.conf. Also create a temporary modulepath.
+# Additionally include rundir, which gets its own initialization.
+puppet_root = Dir.mktmpdir
+moduledir = File.join(puppet_root, 'modules')
+Dir.mkdir(moduledir)
+
+# Use isolated directories in puppet_root *except* confdir so we can connect to puppetserver
+setting_keys = Puppet::Settings::REQUIRED_APP_SETTINGS - [:confdir] + [:rundir]
+cli = (setting_keys).flat_map do |setting|
+  ["--#{setting}", File.join(puppet_root, setting.to_s.chomp('dir'))]
+end
+cli << '--modulepath' << moduledir
+Puppet.initialize_settings(cli)
+
+# Avoid extraneous output
+Puppet[:report] = false
+
+# Make sure to apply the catalog, TODO (on orch side): pipe through _noop
+Puppet[:noop] = apply_opts['noop'] || false
+# TODO(on orch side): enable apply setting, currently this is only configurable from bolt.yaml in bolt
+if apply_opts['apply_settings']
+  apply_opts['apply_settings'].each do |setting, value|
+    Puppet[setting.to_sym] = value
+  end
+end
+
+exit_code = 0
+begin
+
+  # TODO: REMOVE THESE THEY ARE DEBUG LOGS
+  # Puppet::Util::Log.newdestination(:console)
+  # Puppet[:log_level] = :info
+
+  # Sync plugins to cache. Note that we are explicitly *not* using an isolated confdir
+  # because we need to be able to communicate with puppetserver 
+  plugin_cache = args['plugin_cache']
+  plugin_dest = File.join(plugin_cache, 'plugins')
+  pluginfact_dest = File.join(plugin_cache, 'pluginfacts')
+  Puppet[:plugindest] = plugin_dest
+  Puppet[:pluginfactdest] = pluginfact_dest
```

Comment:
Whenever possible, we should be adding settings to the cli - that way any derived settings or implicit early init does the right thing

Reply:
Updated every setting to prefer CLI over settings where possible.